doc/pipeline.md

```diff
@@ -1,51 +1,129 @@
 # Pipeline pattern
 
-The **pipeline** pattern receives a data stream that is processed in several stages. Each stage manages the data produced by the previous stage and delivers its result to the next stage.
+The **pipeline** pattern receives a data stream that is processed in several
+stages. Each stage processes the data produced by the previous stage and delivers
+its result to the next stage.
 
-The interface to the **pipeline** pattern is provided by function `grppi::pipeline()`. As all functions in *GrPPI*, this function takes as its first argument an execution policy.
+The interface to the **pipeline** pattern is provided by function
+`grppi::pipeline()`. As all functions in *GrPPI*, this function takes as its
+first argument an execution policy.
 
 ~~~{.cpp}
 grppi::pipeline(exec, other_arguments...);
 ~~~
 
 ## Pipeline variants
 
-TODO
+There are several variants:
+
+* *Standalone pipeline*: Is a top level pipeline. Invoking the algorithm runs the
+pipeline.
+* *Composable pipeline*: Builds a pipeline object that can be later inserted
+into another pattern.
 
 ## Key elements in pipeline
 
-TODO
+The key elements in a **pipeline** are the **Generator** producing data items
+and the **Transformer** stages.
+
+A **Generator** is any C++ callable entity that takes zero arguments and
+produces data items from a given type. Thus, a **Generator** `gen` is any
```

Comment:
In fact the requirements on type U should be:

a) Type U can be dereferenced through operator*

~~~c++
U item{generator()}
value = *item;
~~~

b) Type U is convertible to bool
~~~c++
U item{generator()};
if (item) { ... } // item holds a value
if (!item) { ... } // item does not hold a value

Please, double check this is right.